### ISSUE TITLE
Failure when using parse_deps to include type from crate dependency

### ISSUE THREAD

**Issue by baumanj** [eqrion/cbindgen#678]
In the [`mp4parse-rust` repo](https://github.com/mozilla/mp4parse-rust/), the core functionality is in the `mp4parse` crate, and `mp4parse_capi` depends upon it to provide the C API. Wanting to add a type to configure the behavior of `m4parse` requires a new type which can be accessed from both crates as well as the FFI caller. I tried to achieve this with one type defined in `mp4parse` and imported into `mp4parse_capi`, but ran into some confusing failures.

Based on this line from [the docs](https://github.com/eqrion/cbindgen/blob/54dfcf89e419afbf60f123088462ef91edd85916/docs.md#writing-your-c-api):
> …it will also spider through your crate (and optionally its dependencies) to try to find the definitions of every type used in your public API

…I tried defining a the new public type `ParseStrictness` in `m4parse` and adding it to the public interface of the FFI function [`mp4parse_avif_new`](https://github.com/mozilla/mp4parse-rust/blob/1ab30201ad06389d0d0ebe3036418d624e7219f4/mp4parse_capi/src/lib.rs#L581) defined in `mp4parse_capi`, but when I ran
```
$ cbindgen --config mp4parse_capi/cbindgen.toml --crate mp4parse_capi
```
…I got the error
```
WARN: Can't find ParseStrictness. This usually means that this type was incompatible or not found.
```
Adding
```toml
[export]
include = ["ParseStrictness"]
```
or
```toml
[export]
include = ["mp4parse::ParseStrictness"]
```
to cbindgen.toml didn't made a difference. Nor did adding `--lockfile Cargo.lock` to the invocation.

I tried forcing cbindgen to parse the `mp4parse` crate by adding:
```toml
[parse]
parse_deps = true
```
which resulted in
```
WARN: Parsing crate `env_logger`: can't find mod imp`.
WARN: Parsing crate `env_logger`: can't find mod imp`.
WARN: Skip aho-corasick::MIN_SKIPS - (not `pub`).
WARN: Skip lazy_static::INIT - (Unsupported expression. Call(ExprCall { attrs: [], func: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident(Lazy), arguments: None }] } }), paren_token: Paren, args: [Call(ExprCall { attrs: [], func: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident(Cell), arguments: None }, Colon2, PathSegment { ident: Ident(new), arguments: None }] } }), paren_token: Paren, args: [Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident(None), arguments: None }] } })] }), Comma, Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident(ONCE_INIT), arguments: None }] } })] }))
WARN: Skip lazy_static::INIT - (Unsupported expression. Call(ExprCall { attrs: [], func: Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident(Lazy), arguments: None }] } }), paren_token: Paren, args: [Path(ExprPath { attrs: [], qself: None, path: Path { leading_colon: None, segments: [PathSegment { ident: Ident(Once), arguments: None }, Colon2, PathSegment { ident: Ident(INIT), arguments: None }] } })] }))
ERROR: Parsing crate `hashbrown`: cannot open file `/Users/jbauman/.cargo/registry/src/github.com-1ecc6299db9ec823/hashbrown-0.9.1/src/mod.rs`.
ERROR: Couldn't generate bindings for /Users/jbauman/src/mp4parse-rust_m-u-current.
```
Then I tried to get it to *only* parse `mp4parse` and no other dependencies with
```toml
[parse]
parse_deps = true
include = ["mp4parse"]
```
which resulted in a crash:
```
WARN: Skip mp4parse::MIN_SIZE - (not `pub`).
WARN: Skip mp4parse::CONFIG_OBUS_OFFSET - (not `pub`).
WARN: Skip mp4parse::MAX - (not `pub`).thread 'main' has overflowed its stack
fatal runtime error: stack overflow
```
and adding `export RUST_BACKTRACE=1` doesn't yield any additional information.

I've tried with cbindgen 0.16.0, 0.18.0 and built from source on [ccea33e](https://github.com/eqrion/cbindgen/tree/ccea33ecb457339a5567a777aa634d87537700d8). All this on macOS Catalina 10.15.7 (19H15) with rust 1.51.0.

**Comment by emilio** [comment]
So #679 is a fix for the hashbrown error. The good thing is that the stack overflow doesn't happen with just `parse_deps = true` after that.

However it still happens with `include = ["mp4parse"]` because without parsing `fallible_collections` cbindgen is fooled into thinking it has a typedef to itself:

A reduced test-case for that is:

```rust
pub type TryVec<T> = fallible_collections::TryVec<T>;
pub type TryString = fallible_collections::TryVec<u8>;
```

When ran with `cbindgen --lang C t.rs`.

**Comment by emilio** [comment]
https://github.com/eqrion/cbindgen/pull/680 fixes the stack overflow too.

**Comment by emilio** [comment]
Should be fixed by the two PRs mentioned above. Let me know if you hit other issues, thanks!

**Comment by baumanj** [comment]
This is great! With the fix, I'm able to succeed with
```toml
[parse]
parse_deps = true
include = ["mp4parse"]
```
For the sake of completeness, I also tried
```toml
[parse]
parse_deps = true
```
But I got failures on the `cast` and `clap` crates, which I filed as https://github.com/eqrion/cbindgen/issues/681. If I exclude those two:
```toml
[parse]
parse_deps = true
exclude = ["cast", "clap"]
```
I get a different failure ([full output here](https://github.com/eqrion/cbindgen/files/6273957/repro.txt)):
```
thread 'main' panicked at 'Iter has 1 params but is being instantiated with 2 values', src/bindgen/ir/opaque.rs:114:9
```
which can be resolved by adding "serde_json" to the exclude list. I couldn't reproduce that exactly by just parsing the crate, and I didn't want to fall too far down the rabbit hole, so I didn't file an issue for that one. But, if you have any difficulty reproducing it, let me know.
